File: src/get_tags.js
const ffprobePath = require('@ffprobe-installer/ffprobe').path;
const { ffprobe } = require('kiss-ffmpeg');
const fs = require('fs');
const mime = require('mime-types');
const path = require('path');
var sizeOf = require('image-size');

ffprobe.command = ffprobePath;

function sleepSync(ms) {
    let start = new Date().getTime(), expire = start + ms;
    while (new Date().getTime() < expire) { }
    return;
}

/*// makes directories synchronously iterable. thankz stack over flow :D
const p = fs.Dir.prototype;
if (p.hasOwnProperty(Symbol.iterator)) { return; }
const entriesSync = function* () {
try {
	let dirent;
	while ((dirent = this.readSync()) !== null) { yield dirent; }
    } finally { this.closeSync(); }
}
if (!p.hasOwnProperty(entriesSync)) { p.entriesSync = entriesSync; }
Object.defineProperty(p, Symbol.iterator, {
	configurable: true,
	enumerable: false,
	value: entriesSync,
	writable: true
});*/

function overallTrackNumber(track, disc, discTracks) {
	let n = track;
	for (let i = 1; i <= disc-1; i++) {
		n += discTracks[i];
	}
	return n;
}

module.exports = async function getTags(form, progressBar) {
	// indeterminate progress bar
	await progressBar.makeIndeterminate();
	await progressBar.setLabel('validating album path..')

	let stats;
	try { // check if form.albumDirectory exists + is a directory
		stats = fs.statSync(form.albumDirectory);
	} catch (err) {
		await progressBar.error(err.toString()); 
		return false; 
	}
	if (stats.isDirectory()) { /* pass */ } else if (stats.isFile()) {
		form.albumDirectory = path.dirname(form.albumDirectory);
	} else {
		await progressBar.error(form.albumDirectory + ' is not a file or directory');
		return false;
	}

	// loop through form.albumDirectory
	await progressBar.setLabel('reading ' + form.albumDirectory + '..');
	const albumDir = fs.opendirSync(form.albumDirectory);
	var audioFiles = [];
	let imageFiles = {};
	let discTracks = {};
	let disc;
	let track;
	let fileMimetype;
	let metadata;
    let promises = [];
	for await (const f of albumDir) { // need to somehow ".then()" this for loop :(
		if (f.name != "concat.wav") {
			fullpath = path.join(form.albumDirectory, f.name);
			await progressBar.setLabel('reading ' + fullpath + '..');

			fileMimetype = mime.lookup(fullpath);
            console.log(fullpath);
			switch (fileMimetype.split("/")[0]) {
				case "audio": // if mimetype is audio/* then get its tags + store in dictionary with path as key
                    promises.push(ffprobe(fullpath).then(function(info) {
                        metadata = info.format;
                        if (metadata.tags.disc) {
						    disc = parseInt(metadata.tags.disc.split("/")[0]);
					    } else { disc = 1; }
					    track = parseInt(metadata.tags.track.split("/")[0]);
					    audioFiles.push({
						    filename: metadata.filename,
						    artist: metadata.tags.artist,
						    albumArtist: metadata.tags.albumArtist || "",
						    title: metadata.tags.title,
						    track: track,
						    disc: disc,
						    time: parseFloat(metadata.duration) // - parseFloat(metadata.start_time)
					    });
					    if (discTracks[disc]) {
						    if (discTracks[disc] < track) {
							    discTracks[disc] = track;
						    }
					    } else {
						    discTracks[disc] = track;
					    }
                    }).catch(function(err) {
                        console.log(err);
                        (async () => {await progressBar.error(err.toString())})();
                    }));
					break;
				case "image": // if mimetype is image/* and form.detectCover add it to cover art candidates list
					if (form.detectCover) imageFiles[f.name] = fullpath;
					break;
				default:
					break;
			}
		}
	}

	await progressBar.setLabel('checking if ' + form.albumDirectory + 'has no sound files..');
	if (!audioFiles) { // if form.albumDirectory has no sound files return an error
		await progressBar.error(form.albumDirectory + ' does not contain any sound files');
		return false;
	}
    
    await progressBar.setLabel('ordering audio files..');
    Promise.all(promises).then(function() { // FURIOUS that i have to put this stupid while loop in here
        let limit = 300;
        let c = 0;
        while (promises.length != audioFiles.length) {
            sleepSync(16);
            console.log("slept");
            c += 1;
            if (c >= limit) {
                console.log("broken");
                break;
            }
        }
        return c < limit;
    }).then(function(ok) {
        if (!ok) {
            Promise.reject('audio files miscounted');
        }
        console.log("VERIFY THAT THERE ARE " + promises.length + " TRACKS IN THE ALBUM");
        console.log(discTracks);
	    audioFiles.sort(function(a, b) {
		    aOverall = overallTrackNumber(a.track, a.disc || 1, discTracks);
		    bOverall = overallTrackNumber(b.track, b.disc || 1, discTracks);
            // console.log(a.track, b.track);
            // console.log(a.disc, b.disc);
            // console.log(aOverall, bOverall);
            // console.log("---------");
            if (aOverall > bOverall) return 1;
            if (aOverall < bOverall) return -1;
            if (aOverall == bOverall) return 0;
	    });
    });

	await progressBar.setLabel('checking cover art..');
	// if form.detectCover is on + no image files set the cover to a 1920 x 1080 all black image
	if (!imageFiles && form.detectCover) { 
		form.coverPath = '../assets/black.png'
	}

	if (!form.detectCover) { // if form.detectCover is off then check if form.coverPath exists
		let stats;
		try { // check if form.coverPath exists + is a file
			stats = fs.statSync(form.coverPath);
		} catch (err) { // if it doesn't exist return an error
			await progressBar.error(err.toString()); 
			return false; 
		}
		if (stats.isFile()) { /* pass */ } else { // if it's not a file return an error
			await progressBar.error(form.coverPath + ' is not a file');
			return false;
		}
	} else {
		await progressBar.setLabel('choosing cover art..');
		let found = false; // check for these names, highest to lowest priority
		["folder.png", "cover.png", "folder.jpg", "cover.jpg"].forEach(function(n) {
			if (n in imageFiles) {
				form.coverPath = imageFiles[n];
				found = true;
			}
		});

		// i could definitely optimize this section what with the recalculating of image sizes 3 times 
		// in a row BUT this is the easiest way for me to conceptualize this and i just want it to work
		if (!found) { // largest image that has "cover" in its name
			let withCover = {};
			let dim;
			Object.keys(imageFiles).forEach(function(key) {
				if (key.toLowerCase().includes("cover")) {
					dim = sizeOf(imageFiles[key]);
					withCover[dim.width*dim.height] = imageFiles[key];
				}
			});
			if (withCover) {
				form.coverPath = withCover[Math.max(...Object.keys(withCover))];
				found = true;
			} 
		}

		if (!found) { // largest image that has "art" in its name
			let withArt = {};
			let dim;
			Object.keys(imageFiles).forEach(function(key) {
				if (key.toLowerCase().includes("art")) {
					dim = sizeOf(imageFiles[key]);
					withCover[dim.width*dim.height] = imageFiles[key];
				}
			});
			if (withArt) {
				form.coverPath = withArt[Math.max(...Object.keys(withArt))];
				found = true;
			}
		}

		if (!found) { // largest image in directory
			let covers = {};
			let dim;
			Object.keys(imageFiles).forEach(function(key) {
				dim = sizeOf(imageFiles[key]);
				covers[dim.width*dim.height] = imageFiles[key];
			});
			form.coverPath = covers[Math.max(...Object.keys(covers))];
		}
	}
	
    await progressBar.setLabel('wait..');
	return { // return object with relevant tags + location of cover art
		form: form,
		audioFiles: audioFiles,
        progressBar: progressBar,
        discTracks: discTracks
	}
};

File: src/get_tags.go
package main

import (
    "errors"
    "os"
    "path"
    "regexp" 
    "sort"
    "strconv"   
    "strings"
    "time"

    "github.com/Akumzy/ipc"
    "github.com/dhowden/tag"
    "github.com/gabriel-vasile/mimetype"
    ffmpeg "github.com/modfy/fluent-ffmpeg"
)

func getTags(channel *ipc.IPC, formData FormData, ffprobePath string) VideoData {
    ffmpeg.SetFfProbePath(ffprobePath)
    validatePaths(channel, formData)

    setLabel(channel, "reading " + path.Base(formData.albumDirectory) + "..")
    albumDirectoryFile, err := os.Open(formData.albumDirectory); if err != nil { panic(err) }
    files, err := albumDirectoryFile.Readdirnames(-1); if err != nil { panic(err) }
    
    discTracks := make(map[uint64]uint64)
    audioFiles := []AudioFile{}
    imageFiles := []string{}

    // wrote this myself :D i'll probably have to change it sometime
    trackRe := regexp.MustCompile(`^([0-9]+|[A-Za-z]{1,2}|[0-9]+[A-Za-z]|)(-| - |_| |)([0-9]+|[A-Za-z])(. | |_)`)

    for _, base := range files {
        setLabel(channel, "reading " + base + "..")
        file := path.Join(formData.albumDirectory, base)
        mime, err := mimetype.DetectFile(file); if err != nil { panic(err) }

        switch strings.Split(mime.String(), "/")[0] {
        case "audio":
            var disc, track uint64
            var artist, albumArtist, title, album string
            ffprobeData, err := ffmpeg.Probe(file); if err != nil { panic(err) }
            metadata := getMetadata(file)
            
            if artist = metadata.Artist(); artist == "" {
                artist = "[unknown artist]"
            }
            
            if albumArtist = metadata.AlbumArtist(); albumArtist == "" {
                albumArtist = "[unknown artist]"
            }

            if title = metadata.Title(); title == "" {
                title = "[untitled]"
            }

            if album = metadata.Album(); album == "" {
                album = "[untitled]"
            }

            year := strconv.Itoa(metadata.Year())
            cover := metadata.Picture()

            seconds, err := strconv.ParseFloat(ffprobeData["format"].(map[string]interface{})["duration"].(string), 64); if err != nil { panic(err) }

            // frankly i dont trust the tag library's assesment of track / disc numbers SORRY
            if (ffprobeData["format"].(map[string]interface{})["tags"].(map[string]interface{})["disc"] != nil) {
                disc = parseTrackTag(ffprobeData["format"].(map[string]interface{})["tags"].(map[string]interface{})["disc"].(string))
            } else { disc = 1 }

            if (ffprobeData["format"].(map[string]interface{})["tags"].(map[string]interface{})["track"] != nil) {
                track = parseTrackTag(ffprobeData["format"].(map[string]interface{})["tags"].(map[string]interface{})["track"].(string))
            } else {
                if !trackRe.MatchString(file) {
                    panic(errors.New("please make sure your filenames start with a track number" +
                        "if they are not tagged properly (which would be preferrable). for exact" + 
                        "specifications as to what does and does not get detected as a track" + 
                        "number see https://github.com/sunglasseds/album2video"))
                }
                track, disc = parseTrack(file, trackRe)
            }
            
            if dt, ok := discTracks[disc]; ok {
                if dt < track {
                    discTracks[disc] = track                
                }
            } else {
                discTracks[disc] = track
            }

            audioFiles = append(audioFiles, AudioFile{
                filename: file,
                artist: artist,
                albumArtist: albumArtist,
                title: title,
                album: album,
                year: year,
                track: track,
                disc: disc,
                time: time.Duration(seconds * float64(time.Second)),
                cover: cover,
                discTracks: &discTracks,
            })

            case "image":
                imageFiles = append(imageFiles, file)
        }
    }    

    setLabel(channel, "ordering audio files..")
    sort.Sort(byTrack(audioFiles))

    Println(channel, audioFiles)

    return VideoData{
        formData: formData,
        audioFiles: audioFiles,
        imageFiles: imageFiles,
        discTracks: discTracks,
    }
}

func validatePaths(channel *ipc.IPC, formData FormData) {
    setLabel(channel, "validating album path..")
    stats, err := os.Stat(formData.albumDirectory); if err != nil { panic(err) }

    if stats.Mode().IsRegular() {
        formData.albumDirectory = path.Dir(formData.albumDirectory)
    } else if stats.IsDir() { /* pass */ } else {
        panic(errors.New("album directory is not a file or directory"))
    }

    if !formData.detectCover {
        setLabel(channel, "validating cover path..")
        stats, err = os.Stat(formData.coverPath); if err != nil { panic(err) }

        if stats.Mode().IsRegular() {
            formData.albumDirectory = path.Dir(formData.albumDirectory)
        } else {
            panic(errors.New(formData.coverPath + " is not a file"))
        }
    }

    setLabel(channel, "validating output path..")
    stats, err = os.Stat(formData.outputPath); if err != nil { panic(err) }

    if formData.separateFiles && stats.Mode().IsRegular() {
        formData.outputPath = path.Dir(formData.outputPath)
    } else if !formData.separateFiles && stats.IsDir() {
        formData.outputPath = path.Join(formData.outputPath, "out.mp4")
    } else if formData.separateFiles == stats.IsDir() { /* pass */ } else {
        panic(errors.New(formData.outputPath + " is not a file or directory"))
    }
}

func getMetadata(filename string) (tag.Metadata) {
    f, err := os.Open(filename); if err != nil { panic(err) }
    m, err := tag.ReadFrom(f); if err != nil { panic(err) }
    err = f.Close(); if err != nil { panic(err) }
    return m
}

func parseTrackTag(v string) (t uint64) {
    t, err := strconv.ParseUint(strings.Split(v, "/")[0], 10, 64)
    if err != nil { panic(err) }; return
}

func parseTrack(filename string, trackRe *regexp.Regexp) (d, t uint64) {
    submatches := trackRe.FindSubmatch([]byte(filename))
    discSubmatch := string(submatches[1])
    trackSubmatch := string(submatches[3])
    
    d, err := strconv.ParseUint(discSubmatch, 10, 64)
    if err != nil {
        d = 1
        switch len(discSubmatch) {
        case 1:
            d += uint64(strings.Index(strings.ToLower(string(discSubmatch[0])), "abcdefghijklmnopqrstuvwxyz"))
            fallthrough
        case 2:
            d += uint64(strings.Index(strings.ToLower(string(discSubmatch[1])), "abcdefghijklmnopqrstuvwxyz")*26)
        default:
            panic(err)
        }
    }

    t, err = strconv.ParseUint(trackSubmatch, 10, 64)
    if err != nil {
        if len(trackSubmatch) != 1 {
            panic(err)
        }
        t = uint64(strings.Index(strings.ToLower(string(trackSubmatch[0])), "abcdefghijklmnopqrstuvwxyz") + 1)
    }
    return
}

// le []AudioFile sorting interface
type byTrack []AudioFile

func (s byTrack) Len() int {
    return len(s)
}

func (s byTrack) Swap(i, j int) {
    s[i], s[j] = s[j], s[i]
}

func (s byTrack) Less(i, j int) bool {
    discTracks := s[i].discTracks
    iOverall := overallTrackNumber(s[i].track, s[i].disc, discTracks)
    jOverall := overallTrackNumber(s[j].track, s[j].disc, discTracks)
    return iOverall < jOverall
}

func overallTrackNumber(track, disc uint64, discTracks *map[uint64]uint64) (n uint64) {
    n = track
    for i := uint64(1); i <= disc-1; i++ {
        n += (*discTracks)[i]
    }
    return n
}
